Can't invoke a 'where F: FnOnce' closure

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
pub enum Option<T> {
    Some(T),
    None,
}

pub use Option::{Some, None};

#[lang = "fn_once"]
pub trait FnOnce<Args> {
    #[lang = "fn_once_output"]
    type Output;

    extern "rust-call" fn call_once(self, args: Args) -> Self::Output;
}

impl<T> Option<T> {
    pub fn map<R, F>(self, f: F) -> Option<R> where F: FnOnce(T) -> R {
        match self {
            Some(value) => Some(f(value)),
            None => None,
        }
    }
}
```

I expected to see this happen: for this to compile just fine

Instead, this happened:

```
crash2.rs:19:33: error: Failed to resolve expression of function call
   19 |             Some(value) => Some(f(value)),
      |                                 ^
crash2.rs:19:33: error: failed to type resolve expression
crash2.rs:19:33: error: failed to resolve argument type
crash2.rs:18:9: error: expected ‘<tyty::error>’ got ‘Option<T?>’
   18 |         match self {
      |         ^~~~~
   19 |             Some(value) => Some(f(value)),
      |                            ~~~~
   20 |             None => None,
      |                     ~~~~
```

This works differently if the `FnOnce` bound is specified inline and not with `FnOnce`, see https://github.com/Rust-GCC/gccrs/issues/2105. Maybe `where` bounds don't make it into `get_specified_bounds ()`?

https://github.com/Rust-GCC/gccrs/blob/92b137048583c2c22c585097119dba6a5fe06228/gcc/rust/typecheck/rust-hir-type-check-expr.cc#L1766

### Meta

- What version of Rust GCC were you using, git sha if possible. 92b137048583c2c22c585097119dba6a5fe06228

--- a/gcc/rust/hir/rust-ast-lower-implitem.h
+++ b/gcc/rust/hir/rust-ast-lower-implitem.h
@@ -133,6 +133,14 @@ public:
   {
     // ignore for now and leave empty
     std::vector<std::unique_ptr<HIR::WhereClauseItem> > where_clause_items;
+    for (auto &item : function.get_where_clause ().get_items ())
+      {
+	HIR::WhereClauseItem *i
+	  = ASTLowerWhereClauseItem::translate (*item.get ());
+	where_clause_items.push_back (
+	  std::unique_ptr<HIR::WhereClauseItem> (i));
+      }
+
     HIR::WhereClause where_clause (std::move (where_clause_items));
     HIR::FunctionQualifiers qualifiers
       = lower_qualifiers (function.get_qualifiers ());
@@ -209,6 +217,14 @@ public:
   {
     // ignore for now and leave empty
     std::vector<std::unique_ptr<HIR::WhereClauseItem> > where_clause_items;
+    for (auto &item : method.get_where_clause ().get_items ())
+      {
+	HIR::WhereClauseItem *i
+	  = ASTLowerWhereClauseItem::translate (*item.get ());
+	where_clause_items.push_back (
+	  std::unique_ptr<HIR::WhereClauseItem> (i));
+      }
+
     HIR::WhereClause where_clause (std::move (where_clause_items));
     HIR::FunctionQualifiers qualifiers
       = lower_qualifiers (method.get_qualifiers ());

--- a/gcc/rust/hir/tree/rust-hir-item.h
+++ b/gcc/rust/hir/tree/rust-hir-item.h
@@ -273,6 +273,8 @@ public:
   TypeBoundWhereClauseItem &operator= (TypeBoundWhereClauseItem &&other)
     = default;
 
+  Location get_locus () const { return locus; }
+
   std::string as_string () const override;
 
   void accept_vis (HIRFullVisitor &vis) override;

--- a/gcc/testsuite/rust/compile/issue-1524.rs
+++ b/gcc/testsuite/rust/compile/issue-1524.rs
@@ -1,4 +1,4 @@
-// { dg-additional-options "-w" }
+// { dg-additional-options "-w -frust-compile-until=nameresolution" }
 // https://github.com/Rust-GCC/gccrs/issues/1524
 // https://github.com/rust-lang/rust/blob/673d0db5e393e9c64897005b470bfeb6d5aec61b/src/test/ui/methods/method-normalize-bounds-issue-20604.rs
 trait Hasher {

--- a/gcc/testsuite/rust/compile/issue-2106.rs
+++ b/gcc/testsuite/rust/compile/issue-2106.rs
@@ -1,0 +1,15 @@
+struct Foo(u32);
+// { dg-warning "struct is never constructed" "" { target *-*-* } .-1 }
+
+pub trait Bar {
+    fn bar(self);
+}
+
+impl Foo {
+    pub fn map<F>(f: F)
+    where
+        F: Bar,
+    {
+        f.bar();
+    }
+}

